public/app/features/scenes/core/SceneDataProviderNode.tsx

```diff
@@ -0,0 +1,155 @@
+  timeRange: SceneTimeRange;
+} & {
+  [key: string]: SceneObject;
+};
+
+export interface SceneDataProviderNodeState extends SceneObjectStatePlain, SceneLayoutState, SceneDataState {
+  queries: DataQueryExtended[];
+  inputParams: SceneDataProviderInputParams;
+}
+
+export class SceneDataProviderNode extends SceneDataObject<SceneDataProviderNodeState> {
+  static Component = SceneDataProviderRenderer;
+  private runner: SceneQueryRunner = new SceneQueryRunner({ queries: [] });
+  private querySub?: Unsubscribable;
+  private staticQueries: DataQueryExtended[] = [];
+
+  private buildStaticQueries() {
+    const staticQueries: DataQueryExtended[] = [];
+
+    for (let i = 0; i < this.state.queries.length; i++) {
+      const query = this.state.queries[i];
+      const staticQuery = cloneDeep(query);
+
+      for (let prop in query) {
+        if (Object.prototype.hasOwnProperty.call(query, prop)) {
+          console.log(prop, query[prop]);
+          if (query[prop] instanceof SceneObjectBase) {
+            const node = query[prop];
+            console.log('updating', prop, node.state.value);
+            staticQuery[prop] = node.state.value;
+          }
+        }
+      }
+
+      staticQueries.push(staticQuery);
+    }
+
+    return staticQueries;
+  }
+
+  activate(): void {
+    super.activate();
+    this.staticQueries = this.buildStaticQueries();
+    this.runner.updateQueries(this.staticQueries);
+
+    const timeRange = this.state.inputParams.timeRange;
+
+    this.subs.add(
+      timeRange.subscribe({
+        next: (timeRange) => {
+          this.runner.runWithTimeRange(timeRange.range).then((data) => {
+            if (!data) {
+              return;
+            }
+
+            if (this.querySub) {
+              this.querySub.unsubscribe();
+            }
+
+            this.querySub = data.subscribe({
+              next: (data) => {
+                console.log('set data', data, data.state);
+                this.setState({ $data: data });
+              },
+            });
+          });
+        },
+      })
+    );
+
+    for (const [_, param] of Object.entries(this.state.inputParams)) {
+      // skip known params
+      if (param === timeRange) {
+        continue;
+      }
+
+      this.subs.add(
+        param.subscribe({
+          next: (value) => {
+            this.staticQueries = this.buildStaticQueries();
+            this.runner.updateQueries(this.staticQueries);
+            this.runner.runWithTimeRange(timeRange.state.range).then((data) => {
+              if (!data) {
+                return;
+              }
+
+              if (this.querySub) {
+                this.querySub.unsubscribe();
+              }
+
+              this.querySub = data.subscribe({
+                next: (data) => {
+                  console.log('set data', data, data.state);
+                  this.setState({ $data: data });
+                },
+              });
+            });
+          },
+        })
+      );
+    }
```

Comment:
Since queries can contain references to `SceneObject`, this makes sure that every SceneObject dependency is subscribed to for queries to be re-run on dependency change.